### PR TITLE
Improve representation of Sorting in memory

### PR DESCRIPTION
**Pull request by samuelgarcia**
This should improve the in memory representation of Sorting using spike vector:
  * faster for multiprocessing
  * faster saving/loading when serialized
  * slower acces to individual to individuala spike train but caching system to compensate

The main important concept is this one:
```python

# this can sometimes have a very slow init (because check, or searchsorted, or mask or dict creation)
sorting = si.read_a_sorting_format()

# we can trnasform to NumpySorting with to_numpy_sorting()
# this have a cost of reading all spikes once.
# but then this sorting is very fast at init and have internally one big vector only
# so easy dumpability
sorting_np = sorting.to_numpy_sorting()

# This is the same but with a sharedmemm buffer
sorting_shm = sorting.to_shared_memory_sorting()

# or also
sorting_np = sorting.to_multiprocessing(n_jobs=1)
sorting_shm = sorting.to_multiprocessing(n_jobs=2)

# sorting_np and sorting_shm are dumpable
d = sorting_shm.to_dict()
d = sorting_np.to_dict()
```

Step 1 :
* [X] Refactor NumpySorting : use internal numpy spike_vector.
* [X] NumpySorting.from_dict() > NumpySorting.from_unit_dict() because was overwriting base.
* [x] deprecate Sorting.get_all_spike_trains() and converge to Sorting.to_spike_vector()
        @alejoe91 @h-mayorquin : this sub step was tedious, please double check that it is OK

Step 2
* [x] implement caching for spike vector
* [x] implement caching for dict of spikes times

Step 3 : 
* [x] Implement sharememNumpySorting
* [x] Add Sorting.to_sharemem()
* [x] Propagate when sorting is used in multiprocessing

Step 4
* [x] Implement new format based on memmap spike vector
* [x] Use this format by default in sorting.save()

This PR also:
  * make numpy sorting dumpable

To be discussed:
  * waveform extension when slicing use a per segment vector slice. I think It should easier to have a global slicing on 
     the full spike_vector. Maybe lets keep in actual state but lets have this in mind for postprocessing refactoring.

### CODE REVIEW

**Comment by h-mayorquin** [comment]
Hey, @samuelgarcia. Each of the steps seem like a big change, I think that it would be easier to review and test if each was a PR on its own.

File: src/spikeinterface/core/numpyextractors.py
Comment (alejoe91) [reply src/spikeinterface/core/numpyextractors.py:197]: agreed, we should at least deprecate it...

Comment (samuelgarcia) [reply src/spikeinterface/core/numpyextractors.py:201]: No, the to_dict() was a huge mistake because it overwrited the to_dict() to dump the object.
It was use only internally.
So I think it is OK to get it back without transitory process

**Comment by samuelgarcia** [comment]
> Hey, @samuelgarcia. Each of the steps seem like a big change, I think that it would be easier to review and test if each > was a PR on its own.

Of course and we decide with Alessio to make 4 small PRs instead of one big but :
  *b ecause I am lazy
  * because I want to publicaly show that I am stubborn
  * because Iwant to revenge against the pre-commit.ci

then I decided alone to do the exact oposite!
Feel free to hardly critize this choice and hate me for a while.

File: src/spikeinterface/core/numpyextractors.py
Comment (samuelgarcia) [reply src/spikeinterface/core/numpyextractors.py:201]: And also now the NumpySorting is dumpable so we really need the from_dict()/to_dict() to have the true behavior that come from the base.

File: src/spikeinterface/core/numpyextractors.py
Comment (DradeAW) [inline src/spikeinterface/core/numpyextractors.py:348]: Shouldn't it be `SharedMemorySorting`?

Comment (samuelgarcia) [reply src/spikeinterface/core/numpyextractors.py:348]: Oups. thank you. Going back to my english class now.

**Comment by h-mayorquin** [comment]
All right, this will make us it harder for us to review. How are you profiling memory and performance? 

I would like to run some tests on windows and if they are similar to what you are doing it would be better.

**Comment by samuelgarcia** [comment]
Hi Ramon, at the moment this is only concepts. I haven't tested performences.
I will do.

For windows, I do not now how well it will work. For sure the dumpability and having one unique vector will help a lot because otherwie we where picking many many arrays (one per units / segmnts).

I add a `Sorting.to_multiprocessing()` facility. If the sharedmemory machinery do not work well on window we can keep the NumpySorting in that function,  this will be centralized.

**Comment by h-mayorquin** [comment]
@samuelgarcia 
This work for me:
```python
import numpy as np
import json

# Define the structured array
data = np.array([(1, "John", 28),
                 (2, "Alice", 32),
                 (3, "Bob", 45)],
                dtype=[('id', int), ('name', 'U10'), ('age', int)])

# Convert the structured array to a nested list
nested_list = data.tolist()

# Dump the nested list to a JSON file
from spikeinterface.core.core_tools import SIJsonEncoder

json_string = json.dumps(data, cls=SIJsonEncoder)
json.loads(json_string)
```

But most likely re-intialization does not work well when loaded from here if your class is expecting an structured array input. You will ned do something like this:

https://github.com/SpikeInterface/spikeinterface/blob/11a22b7728f0604778659f7536728f155cc433d6/src/spikeinterface/core/numpyextractors.py#L37-L44

File: src/spikeinterface/core/numpyextractors.py
Comment (h-mayorquin) [inline src/spikeinterface/core/numpyextractors.py:141]: This is a good change in naming.

File: src/spikeinterface/core/basesorting.py
Comment (h-mayorquin) [inline src/spikeinterface/core/basesorting.py:218]: I think `npz_folder` and `numpy_folder` are pretty obscure references. Probably it will be a good idea to add clarification of pros and cons in the docstring.

Also, an annoying thing of the pattern we are using right now where extractor.save is called from the base class is that this string will not propagate to that method.

**Review by h-mayorquin — COMMENTED** [review]
Took the first look at it. I will need to read this a couple of times. Surprised you rae using Folder instead of Directory here.  I thought you were a linux-all-in kind of guy : P

File: src/spikeinterface/core/basesorting.py
Comment (h-mayorquin) [reply src/spikeinterface/core/basesorting.py:218]: Short name is fine if it is documented where it is used. right now we don't have documentation of these two formats anywhere do we?

**Comment by samuelgarcia** [comment]
@alejoe91 : it would be really nice to merge this soon after the release because because this make many changes that could lead to conflict tedious to track if we accept more PR in between

File: src/spikeinterface/core/tests/test_basesorting.py
Comment (h-mayorquin) [inline src/spikeinterface/core/tests/test_basesorting.py:172]: To be honest, I am oddly suprised that you decided to test deprecation warnings. I think that I am on the side that is likely to be wrong by making too much testing and this is like on the line for me.

Not that I think is wrong, just pondering about my surprise.

File: src/spikeinterface/core/numpyextractors.py
Comment (h-mayorquin) [inline src/spikeinterface/core/numpyextractors.py:102]: Great that you added a docstring!

File: src/spikeinterface/core/numpyextractors.py
Comment (h-mayorquin) [inline src/spikeinterface/core/numpyextractors.py:107]: convenience*

It would be even better if we added the specific methods that you mentioned. I can add them later after we merge as well.

File: src/spikeinterface/core/numpyextractors.py
Comment (h-mayorquin) [inline src/spikeinterface/core/numpyextractors.py:122]: How funny that I neve realized that the intialization was done always done with static method. 

Very elegant that you fit this right into it. As if this is the right datastructure to initialize this object. I like it.

File: src/spikeinterface/core/numpyextractors.py
Comment (h-mayorquin) [reply src/spikeinterface/core/numpyextractors.py:206]: Yeah, I remember I was confused about this a month ago. This is the right call. It is the current method that is overloaded with the other `to_dict` as you were telling us.

File: src/spikeinterface/core/numpyextractors.py
Comment (h-mayorquin) [inline src/spikeinterface/core/numpyextractors.py:352]: Does this works on windows?

File: src/spikeinterface/core/basesorting.py
Comment (h-mayorquin) [inline src/spikeinterface/core/basesorting.py:413]: I think that `to_spike_vector` can be confusing because we have the sorter methods `to_numpy_sorting`, `to_shared_memory_sorting` and `to_dict` that transform the sorter and output a different representation of the sorter.

`to_spike_vector` though just builds a vector representation of the spikes. I think that something like `build_spike_vector` would be easier to discover when someone is exploring the sorter with auto-completion and will help disambiguate the semantic convention that we have in the `to_something` methods.

**Review by h-mayorquin — COMMENTED** [review]
This is cool, there are some tests failing but I review some of the core functionality as I guess that would not change that much.

I am still thinking about the profiling about as I think from the big changes here:
* Changing NumpySorting to operate with a vector
* Spike vector caching
* SharedMem

This covers the third one? I need to take a second look, this would have been easier if we separated these things in different PRs : P

Overall this looks good, I hope I can help with some of the profiling. Thanks for this work, Sam.

File: src/spikeinterface/core/sortingfolder.py
Comment (h-mayorquin) [reply src/spikeinterface/core/sortingfolder.py:104]: Same in the sorter above.

**Comment by h-mayorquin** [comment]
Something else that I am thinking about.
We have the convention of having `BinaryRecordingExractor` for the bare bones and `BinaryFolderRecording` for loading `BinaryRecordingExractor` with all the metadata and the stuff. The recording object of our internal format.

Now we do the same with `NumpyFolderSorting` which is going to play a similar role as the sorter of our internal format.

When I started exploring the library I found the Folder name confusing in an extractor. I was thinking to myself: why do we have `BinaryFolderRecording` when we already have `BinaryRecordingExractor`, aren't they kind of doing the same? And I think I asked you guys about it.

Now I am thinking you are losing an opportunity for disambiguation branding, @samuelgarcia @alejoe91 !
What if we named this `SpikeInterfaceExtractor` and `SpikeInterfaceSorter` I think this a class name that reflects the intent very clearly and also bears our prideful name.

File: src/spikeinterface/core/basesorting.py
Comment (samuelgarcia) [reply src/spikeinterface/core/basesorting.py:10]: I think global variable is OK.

File: src/spikeinterface/core/basesorting.py
Comment (samuelgarcia) [reply src/spikeinterface/core/basesorting.py:227]: oups I forgot to remove this line

File: src/spikeinterface/core/basesorting.py
Comment (samuelgarcia) [reply src/spikeinterface/core/basesorting.py:299]: In many case the time vector is be already there (numpy, sharemem, ...), so clearly faster.
In other case we need to load all spiketrain anyway so the cost is almost the same.

File: src/spikeinterface/core/basesorting.py
Comment (samuelgarcia) [reply src/spikeinterface/core/basesorting.py:413]: good idea. @alejoe91 are you agree with this naming proposal ?
This could break some old code.

File: src/spikeinterface/core/basesorting.py
Comment (samuelgarcia) [reply src/spikeinterface/core/basesorting.py:415]: concatenated=True/False is truely need. because in many place we want to loop over segments without to cut at the limit.

use_cache is also important in case you want to use it once and do not want to keep it in memory.

extremum_channel is need for spike location computation. This could be removed here and add more complexity on compute_spike_olocation side. Not sure if this is relevant or not.

File: src/spikeinterface/core/basesorting.py
Comment (samuelgarcia) [reply src/spikeinterface/core/basesorting.py:450]: I do not think I understant your question. There are 2 internal caches.
dict of spike train and spike vector.
get_unit_spike_train cache the dict and to_spike_vector cache the spike vector.

File: src/spikeinterface/core/basesorting.py
Comment (samuelgarcia) [reply src/spikeinterface/core/basesorting.py:477]: thanks

File: src/spikeinterface/core/basesorting.py
Comment (samuelgarcia) [reply src/spikeinterface/core/basesorting.py:506]: I am learning from masters.

File: src/spikeinterface/core/numpyextractors.py
Comment (samuelgarcia) [reply src/spikeinterface/core/numpyextractors.py:102]: :)

File: src/spikeinterface/core/tests/test_basesorting.py
Comment (alejoe91) [reply src/spikeinterface/core/tests/test_basesorting.py:172]: @samuelgarcia is changing....

File: src/spikeinterface/core/numpyextractors.py
Comment (samuelgarcia) [reply src/spikeinterface/core/numpyextractors.py:107]: There are already. They are class method. I enhance the doc.

File: src/spikeinterface/core/numpyextractors.py
Comment (samuelgarcia) [reply src/spikeinterface/core/numpyextractors.py:352]: I think this work. We use to have bugs when using sharedmem stuff but we were not cleaning it properly.
We can make more test for this on windows platform.

File: src/spikeinterface/core/sortingfolder.py
Comment (samuelgarcia) [reply src/spikeinterface/core/sortingfolder.py:104]: I think that using "d" for a dict when the piece of code around 5 lines make the reading easier than "dictionary"
I known that your prefer a longer variable name.

File: src/spikeinterface/core/basesorting.py
Comment (alejoe91) [reply src/spikeinterface/core/basesorting.py:412]: yes, sounds good to me

File: src/spikeinterface/core/basesorting.py
Comment (alejoe91) [reply src/spikeinterface/core/basesorting.py:414]: maybe move the new args after `extremum_channel_inds=None`?

File: src/spikeinterface/core/basesorting.py
Comment (alejoe91) [inline src/spikeinterface/core/basesorting.py:511]:
```suggestion
        useful to have it in memory with a unique vector representation.
```

File: src/spikeinterface/core/tests/test_basesorting.py
Comment (alejoe91) [inline src/spikeinterface/core/tests/test_basesorting.py:123]: can you add a test with `save` with more than ` job? This should also run on Windows

File: src/spikeinterface/exporters/to_phy.py
Comment (alejoe91) [inline src/spikeinterface/exporters/to_phy.py:161]: Note that Phy only works for 1 segment

**Comment by alejoe91** [comment]
> Something else that I am thinking about. We have the convention of having `BinaryRecordingExractor` for the bare bones and `BinaryFolderRecording` for loading `BinaryRecordingExractor` with all the metadata and the stuff. The recording object of our internal format.
> 
> Now we do the same with `NumpyFolderSorting` which is going to play a similar role as the sorter of our internal format.
> 
> When I started exploring the library I found the Folder name confusing in an extractor. I was thinking to myself: why do we have `BinaryFolderRecording` when we already have `BinaryRecordingExractor`, aren't they kind of doing the same? And I think I asked you guys about it.
> 
> Now I am thinking you are losing an opportunity for disambiguation branding, @samuelgarcia @alejoe91 ! What if we named this `SpikeInterfaceExtractor` and `SpikeInterfaceSorter` I think this a class name that reflects the intent very clearly and also bears our prideful name. Most people access this through the loading methods so I am not afraid about backwards compatibility.

I agree! We can rename the `Folder` classes as `SpikeInterfaceRecordingFolder`/`SpikeInterfaceSortingFolder`/`SpikeInterfaceSnippetsFolder`. We can also do it in a separate PR. @h-mayorquin could you open an issue about it?

**Comment by samuelgarcia** [comment]
> I agree! We can rename the Folder classes as >SpikeInterfaceRecordingFolder/SpikeInterfaceSortingFolder/SpikeInterfaceSnippetsFolder. We can also do it in a separate > PR. @h-mayorquin could you open an issue about it?

In fact this is not a good idea because we will be able to reload folder from previous version and this will prevent us to make later a new (V3) folder version of recording or sorting.
Lets discuss this but I think this is outside this PR.

File: src/spikeinterface/exporters/to_phy.py
Comment (samuelgarcia) [reply src/spikeinterface/exporters/to_phy.py:161]: I know this make the code more explicit no ?

File: src/spikeinterface/core/tests/test_basesorting.py
Comment (samuelgarcia) [reply src/spikeinterface/core/tests/test_basesorting.py:123]: yep

File: src/spikeinterface/core/basesorting.py
Comment (samuelgarcia) [reply src/spikeinterface/core/basesorting.py:412]: `get_spike_vector()`

Comment (alejoe91) [reply src/spikeinterface/core/basesorting.py:412]: Again, let's keep it for now and change it in a future PR

File: src/spikeinterface/core/tests/test_basesorting.py
Comment (samuelgarcia) [reply src/spikeinterface/core/tests/test_basesorting.py:123]: NumpyFolder and NpzFolder do not have parralel writting.
The shared mem sorting is already tested in many place because we do `to_multiprocessing()` for waveform extractor

**Review by alejoe91 — APPROVED** [review]

**Comment by alejoe91** [comment]
@h-mayorquin @samuelgarcia looks good to me!!

**Comment by h-mayorquin** [comment]
Give me some time to do one last check today in the afternoon.

**Review by h-mayorquin — CHANGES_REQUESTED** [review]
Give me some time to do one last check today in the afternoon.

File: src/spikeinterface/core/basesorting.py
Comment (h-mayorquin) [reply src/spikeinterface/core/basesorting.py:412]: If Sam is fine with `get_spike_vector` I am as well. I think that's an improvement over the existing one that we can release with this big change. If you don't have any reason against I suggest we change it now.

File: src/spikeinterface/core/numpyextractors.py
Comment (h-mayorquin) [reply src/spikeinterface/core/numpyextractors.py:355]: OK.

**Review by h-mayorquin — COMMENTED** [review]
OK, Sam, thanks for waiting for me to have one last look. I added a bunch of comments but there is three things that I would really like to see In their order of importance:

1.  Change of  `get_spike_vector` to `get_spike_vector` if Alessio agrees.
2.  Send `extremum_channel_inds` logic to `compute_spike_olocation`.
3. Use another single letter variable instead of d for dictionaries in the added code.

I did not have time to look at the SharedMemory thing but I guess Alessio did.

**Review by h-mayorquin — COMMENTED** [review]
OK, Sam, thanks for waiting for me to have one last look. I added a bunch of comments but there is three things that I would really like to see In their order of importance:

1.  Change of  `get_spike_vector` to `get_spike_vector` if Alessio agrees.
2.  Send `extremum_channel_inds` logic to `compute_spike_olocation`.
3. Use another single letter variable instead of d for dictionaries in the added code.

I did not have time to look at the SharedMemory thing but I guess Alessio did.

File: src/spikeinterface/core/sortingfolder.py
Comment (h-mayorquin) [reply src/spikeinterface/core/sortingfolder.py:15]: Actually done.

File: src/spikeinterface/core/sortingfolder.py
Comment (h-mayorquin) [reply src/spikeinterface/core/sortingfolder.py:19]: Also, apparently got an old  version

**Comment by samuelgarcia** [comment]
Hi Ramon.
Thank for last feedback.


> Change of get_spike_vector to get_spike_vector if Alessio agrees.

OK for this but later in an other PR.
I will not have time to include this now because it needs more check to see if it do not break deeper and old code.
I really need this PR merge before all other ones and I lease for vaccation in a few day and Alessio will be also out for a while.


> Send extremum_channel_inds logic to compute_spike_olocation.

Same anwser. I agree but lets do this outside this PR. We can open an issue to not forget.


> Use another single letter variable instead of d for dictionaries in the added code.

OK I will do.

> I did not have time to look at the SharedMemory thing but I guess Alessio did.

I think yes.

**Comment by h-mayorquin** [comment]
Ok, @samuelgarcia 

> OK for this but later in an other PR.
I will not have time to include this now because it needs more check to see if it do not break deeper and old code.
I really need this PR merge before all other ones and I lease for vaccation in a few day and Alessio will be also out for a while.

Maybe you can just create a the new method and have the old method call the new one with a deprecation warning? That should be quick, work as a reminder and not break any old code. 

I will open an issue about the rest of the stuff that we discussed here as soon as I have time.

**Comment by DradeAW** [comment]
Could you merge the main branch onto this one?
I wanted to check that Lussac works with this new representation of sorting objects, but it's missing a few commits that make my algorithm crash

**Comment by h-mayorquin** [comment]
@DradeAW done
You can do this locallily btw:

```
git switch name_of_same_remote/sharedmem_sorting
git merge upstream/main 
```

**Comment by DradeAW** [comment]
All tests passed on my end!
I had to change some bits of code because of the new nomenclature, but in general I think the new names are better :)

Awesome stuff!

File: src/spikeinterface/core/sortingfolder.py
Comment (samuelgarcia) [reply src/spikeinterface/core/sortingfolder.py:34]: done

File: src/spikeinterface/core/basesorting.py
Comment (samuelgarcia) [reply src/spikeinterface/core/basesorting.py:219]: The save_kwargs was already there. It is to be flexible with differents format like we do in the rec._save()
Lets keep thos for now.

Comment (h-mayorquin) [reply src/spikeinterface/core/basesorting.py:219]: Ok, true that they were already there.

Let's discuss that in other PR issue as you say.

File: src/spikeinterface/core/basesorting.py
Comment (samuelgarcia) [reply src/spikeinterface/core/basesorting.py:427]: Ok I changed this.

File: src/spikeinterface/core/basesorting.py
Comment (samuelgarcia) [reply src/spikeinterface/core/basesorting.py:432]: Done

**Comment by samuelgarcia** [comment]
@h-mayorquin : I did a few changes and then I merge this if this pass tests.

**Comment by h-mayorquin** [comment]
Sounds good to me, @samuelgarcia .

**Comment by samuelgarcia** [comment]
Thanks a  lot.
@DradeAW : do you see some improvement ?
I think we will have more improvement when we will refactor the waveform extractor object I think.
But on some code it should be already faster to have the cache and no copy for parralel processing.

**Comment by DradeAW** [comment]
> @DradeAW : do you see some improvement ?

It's hard to know for sure, I didn't make reproducible experiments.
For sure it's between 0-10% faster (so at worse the same in my case).

> I think we will have more improvement when we will refactor the waveform extractor object I think.

Can't wait! :)

**Pull request merged**
